dare.go

```diff
@@ -12,194 +12,260 @@
-// all received data. DecryptReader returns an error if the provided decryption
-// configuration is invalid.
-func DecryptReader(src io.Reader, config Config) (io.Reader, error) {
-	if err := setConfigDefaults(&config); err != nil {
-		return nil, err
+func (ad *authDecV10) Open(dst, src []byte) error {
+	header := headerV10(src[:headerSize])
+	if header.Version() != Version10 {
+		return errUnsupportedVersion
+	}
+	if header.Cipher() > CHACHA20_POLY1305 {
+		return errUnsupportedCipher
+	}
+	aeadCipher := ad.Ciphers[header.Cipher()]
+	if aeadCipher == nil {
+		return errUnsupportedCipher
+	}
+	if headerSize+header.Len()+tagSize != len(src) {
+		return errInvalidPayloadSize
+	}
+	if header.SequenceNumber() != ad.SeqNum {
+		return errPackageOutOfOrder
 	}
-	return decryptReaderV10(src, &config)
+	ciphertext := src[headerSize : headerSize+header.Len()+tagSize]
+	if _, err := aeadCipher.Open(dst[:0], header.Nonce(), ciphertext, header.AddData()); err != nil {
+		return errTagMismatch
+	}
+	ad.SeqNum++
+	return nil
 }
 
-// EncryptWriter wraps the given dst and returns an io.WriteCloser which
-// encrypts all data written to it. EncryptWriter returns an error if the
-// provided decryption configuration is invalid.
-//
-// The returned io.WriteCloser must be closed successfully to finalize the
-// encryption process.
-func EncryptWriter(dst io.Writer, config Config) (io.WriteCloser, error) {
-	if err := setConfigDefaults(&config); err != nil {
-		return nil, err
-	}
-	return encryptWriterV10(dst, &config)
+type authEncV20 struct {
+	authEnc
+	finalized bool
 }
 
-// DecryptWriter wraps the given dst and returns an io.WriteCloser which
-// decrypts all data written to it. DecryptWriter returns an error if the
-// provided decryption configuration is invalid.
-//
-// The returned io.WriteCloser must be closed successfully to finalize the
-// decryption process.
-func DecryptWriter(dst io.Writer, config Config) (io.WriteCloser, error) {
-	if err := setConfigDefaults(&config); err != nil {
-		return nil, err
+func newAuthEncV20(cfg *Config) (authEncV20, error) {
+	cipherID := cfg.CipherSuites[0]
+	cipher, err := supportedCiphers[cipherID](cfg.Key)
+	if err != nil {
+		return authEncV20{}, err
+	}
+	var randVal [12]byte
+	if _, err = io.ReadFull(cfg.Rand, randVal[:]); err != nil {
+		return authEncV20{}, err
 	}
-	return decryptWriterV10(dst, &config)
+	return authEncV20{
+		authEnc: authEnc{
+			CipherID: cipherID,
+			RandVal:  randVal[:],
+			Cipher:   cipher,
+			SeqNum:   cfg.SequenceNumber,
+		},
+	}, nil
 }
 
-func defaultCipherSuites() []byte {
-	if hasAESNISupport() {
-		return []byte{AES_256_GCM, CHACHA20_POLY1305}
+func (ae *authEncV20) Seal(dst, src []byte)      { ae.seal(dst, src, false) }
+func (ae *authEncV20) SealFinal(dst, src []byte) { ae.seal(dst, src, true) }
+
+func (ae *authEncV20) seal(dst, src []byte, finalize bool) {
+	if finalize {
+		ae.finalized = true
```

Comment:
should we simply do `ae.finalized = finalize` ?

Reply:
We can do that.

Reply:
Actually I wouldn't do that b/c:
```
Seal(...)
SealFinal(...) // <- finalized == true here
Seal(...) // <- finalized == false here again.
```
While the calling code should ensure that this does not happen - it is something which can be easily overlooked... - For example I did it ^^

Reply:
Added a solution covering both cases - PTAL.